### PR TITLE
Bundles distribution

### PR DESCRIPTION
**Pull request by Mike-S-Mike**
# Checklist:

- [x] Code follows the style guidelines
- [ ] Documentation updated
- [ ] Corner cases evaluated and handled
- [ ] index.js exports updated
- [x] Automatic tests added/updated
- [x] Automatic tests pass
- [ ] Sanity check passed
- [ ] Self-review passed
- [x] Version updated

Implemented DMP for bundles distribution in smart contracts;
Fixed unit tests for smart contracts;
Fixed gas benchmarks;

### CODE REVIEW

File: test/contracts/front/challenges.js [hunk 1]
```diff
@@ -628,24 +674,144 @@ describe('Challenges Contract', () => {
     });
   });
 
-  describe('Calculating cooldown', () => {
-    it('returns 0 if there are less then 4 atlas nodes', async () => {
-      await setNumberOfStakers(2);
-      expect(await getCooldown()).to.equal('0');
+  describe('DMP algorithm testing', () => {
```
Comment (vanruch) [inline test/contracts/front/challenges.js]: I would prefer testing library itself. It would allow us to skip nodes onboarding and make tests much much simplier.

File: contracts/Configuration/Config.sol [hunk 2]
```diff
@@ -16,19 +16,16 @@ contract Config {
...
-    uint32 constant public COOLDOWN_TIMEOUT = 1 days;
 
     uint constant public APOLLO_DEPOSIT = 250000 ether;
-    uint constant public ATLAS1_STAKE = 10000 ether;
-    uint constant public ATLAS2_STAKE = 30000 ether;
-    uint constant public ATLAS3_STAKE = 75000 ether;
+
+    uint[3] public ATLAS_STAKE = [10000 ether, 30000 ether, 75000 ether];
+    uint[3] public ATLAS_NUMERATOR = [1, 2, 7];
+    uint constant public ATLAS_TYPES = 3;
```
[the remark this answers is not in the document]
Comment (Mike-S-Mike) [reply contracts/Configuration/Config.sol]: Fixed

File: contracts/Lib/DMPalgorithm.sol [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+*/
+
+pragma solidity ^0.4.23;
+
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+import "./SafeMathExtensions.sol";
+
+
+library DMPalgorithm {
```
[the remark this answers is not in the document]
Comment (Mike-S-Mike) [reply contracts/Lib/DMPalgorithm.sol]: Fixed

File: contracts/Lib/DMPalgorithm.sol [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            uint currentNum = atlasCount[i].mul(ATLAS_NUMERATOR[i]);
+            if (currentNum == 0) {
+                wcd[i] = 0;
+            }
+            if (currentNum == denominator) {
+                wcd[i] = DMP_PRECISION;
+                currentWCD = DMP_PRECISION;
+            } else {
+                currentNum = currentNum.mul(DMP_PRECISION);
+                currentNum = currentNum.div(denominator);
```
[the remark this answers is not in the document]
Comment (Mike-S-Mike) [reply contracts/Lib/DMPalgorithm.sol]: Fixed

File: contracts/Front/Challenges.sol [hunk 5]
```diff
@@ -152,6 +153,25 @@ contract Challenges is Base {
         return challengerId;
     }
 
+    function getChallengeDesignatedShelterer(bytes32 challengeId) public view returns (address) {
+        uint64 challengeDuration = time.currentTimestamp() - getChallengeCreationTime(challengeId);
```
[the remark this answers is not in the document]
Comment (Mike-S-Mike) [reply contracts/Front/Challenges.sol]: Fixed

File: contracts/Lib/DMPalgorithm.sol [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                wcd[i] = currentWCD;
+            }
+        }
+        wcd[length-1] = DMP_PRECISION;
+
+        uint chosenStake = length - 1;
+        uint randomNumber = uint(DMPbaseHash).mod(DMP_PRECISION);
+        for (i = 0; i < length; i++) {
+            if (wcd[i] != 0 && randomNumber <= wcd[i]) {
+                chosenStake = i;
```
[the remark this answers is not in the document]
Comment (Mike-S-Mike) [reply contracts/Lib/DMPalgorithm.sol]: Fixed

File: test/contracts/front/challenges.js (hunk 1 above)
Comment (Mike-S-Mike) [reply test/contracts/front/challenges.js]: Fixed
I have moved DMP connected unit tests to the separate file. Nevertheless, some tests are still needed for the round selecting functionality, so I have left them in "Designated shelterer" part.

File: test/contracts/lib/dmp_algorithm.js [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+  ATLAS2_NUMENATOR,
+  ATLAS3_NUMENATOR
+} from '../../../src/constants';
+
+chai.use(sinonChai);
+chai.use(chaiAsPromised);
+
+const {expect} = chai;
+
+describe('DMP alrgorithm library', () => {
```
[the remark this answers is not in the document]
Comment (Mike-S-Mike) [reply test/contracts/lib/dmp_algorithm.js]: Fixed

File: test/contracts/lib/dmp_algorithm.js [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+  beforeEach(async () => {
+    snapshotId = await makeSnapshot(web3);
+  });
+
+  afterEach(async () => {
+    await restoreSnapshot(web3, snapshotId);
+  });
+
+  describe('DMP algorithm testing', () => {
```
[the remark this answers is not in the document]
Comment (Mike-S-Mike) [reply test/contracts/lib/dmp_algorithm.js]: Fixed

File: test/contracts/lib/dmp_algorithm.js [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+  let DmpAlgorithmAdapter;
+  let time;
+  let web3;
+  let snapshotId;
+  let context;
+  const now = 1500000000;
+  const challengeId = utils.keccak256('someChallengeId');
+  const sequenceNumber = 2;
+  const creationTime = now;
+  const atlasNums = Array.from([ATLAS1_NUMENATOR, ATLAS2_NUMENATOR, ATLAS3_NUMENATOR]);
```
[the remark this answers is not in the document]
Comment (Mike-S-Mike) [reply test/contracts/lib/dmp_algorithm.js]: Fixed

File: test/contracts/lib/dmp_algorithm.js [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+  const atlasCounts = Array.from([2, 4, 8]);
```
[the remark this answers is not in the document]
Comment (Mike-S-Mike) [reply test/contracts/lib/dmp_algorithm.js]: Fixed

File: test/contracts/lib/dmp_algorithm.js [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+  const atlasCountsNoLast = Array.from([2, 4, 0]);
```
[the remark this answers is not in the document]
Comment (Mike-S-Mike) [reply test/contracts/lib/dmp_algorithm.js]: Fixed

File: src/constants.js [hunk 12]
```diff
@@ -24,3 +24,11 @@ export const ATLAS3_STAKE = utils.toWei('75000');
 export const APOLLO_DEPOSIT = utils.toWei('250000');
 
 export const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';
+
+
+export const DMP_PRECISION = 100;
+export const ROUND_DURATION = 10 * 60;
+export const FIRST_PHASE_DURATION = 2 * 24 * 60 * 60;
+export const ATLAS1_NUMENATOR = 1;
```
[the remark this answers is not in the document]
Comment (Mike-S-Mike) [reply src/constants.js]: Fixed

File: contracts/Lib/DmpAlgorithm.sol [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+    using SafeMathExtensions for uint;
+
+    uint constant public DMP_PRECISION = 100;
+
+
+    function qualifyShelterer(bytes32 dmpBaseHash, uint dmpLength, uint currentRound) internal pure returns (uint32) {
+        return uint32(uint256(keccak256(abi.encodePacked(dmpBaseHash, currentRound))).mod(dmpLength));
+    }
+
+    function qualifyShelterTypeStake(bytes32 dmpBaseHash, uint[] atlasCount, uint[] ATLAS_NUMERATOR, uint32 length) internal pure returns (uint32) {
```
[the remark this answers is not in the document]
Comment (Mike-S-Mike) [reply contracts/Lib/DmpAlgorithm.sol]: Fixed

File: contracts/Lib/DmpAlgorithm.sol [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        uint randomNumber = uint(dmpBaseHash).mod(DMP_PRECISION);
+        for (i = 0; i < length; i++) {
+            if (wcd[i] != 0 && randomNumber <= wcd[i]) {
+                break;
+            }
+        }
+
+        return i;
```
[the remark this answers is not in the document]
Comment (Mike-S-Mike) [reply contracts/Lib/DmpAlgorithm.sol]: Fixed

File: contracts/Lib/DmpAlgorithm.sol [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+    function qualifyShelterer(bytes32 dmpBaseHash, uint dmpLength, uint currentRound) internal pure returns (uint32) {
+        return uint32(uint256(keccak256(abi.encodePacked(dmpBaseHash, currentRound))).mod(dmpLength));
+    }
+
+    function qualifyShelterTypeStake(bytes32 dmpBaseHash, uint[] atlasCount, uint[] ATLAS_NUMERATOR, uint32 length) internal pure returns (uint32) {
+        uint denominator = 0;
+        uint32 i = 0;
+
+        for (i = 0; i < length; i++) {
```
[the remark this answers is not in the document]
Comment (Mike-S-Mike) [reply contracts/Lib/DmpAlgorithm.sol]: Fixed

File: contracts/Lib/DmpAlgorithm.sol [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+            uint elem = atlasCount[i].mul(ATLAS_NUMERATOR[i]);
+            denominator = denominator.add(elem);
+        }
+
+        uint currentWCD = 0;
+        uint[] memory wcd = new uint[](length);
+
+        for (i = 0; i < length; i++) {
```
[the remark this answers is not in the document]
Comment (Mike-S-Mike) [reply contracts/Lib/DmpAlgorithm.sol]: Fixed

File: contracts/Lib/DmpAlgorithm.sol [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        for (i = length - 1; i >= 0; i--) {
+            if (wcd[i] != 0) {
+                wcd[i] = DMP_PRECISION;
+                break;
+            }
+        }
+
+        uint randomNumber = uint(dmpBaseHash).mod(DMP_PRECISION);
```
[the remark this answers is not in the document]
Comment (Mike-S-Mike) [reply contracts/Lib/DmpAlgorithm.sol]: Fixed

File: contracts/Lib/DmpAlgorithm.sol [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+            uint elem = atlasCount[i].mul(ATLAS_NUMERATOR[i]);
+            denominator = denominator.add(elem);
+        }
+
+        uint currentWCD = 0;
+        uint[] memory wcd = new uint[](length);
+
+        for (i = 0; i < length; i++) {
+            uint currentNum = atlasCount[i].mul(ATLAS_NUMERATOR[i]);
+            if (currentNum == 0) {
```
[the remark this answers is not in the document]
Comment (Mike-S-Mike) [reply contracts/Lib/DmpAlgorithm.sol]: Fixed

File: contracts/Lib/DmpAlgorithm.sol [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+
+    function calculateWcd(uint[] atlasCount, uint[] ATLAS_NUMERATOR) private pure returns (uint[]) {
+        uint[] memory wcd = new uint[](ATLAS_NUMERATOR.length);
+        uint currentWCD = 0;
+        uint denominator = calculateDenominator(atlasCount, ATLAS_NUMERATOR);
+
+        for (uint tier = 0; tier < ATLAS_NUMERATOR.length; tier++) {
+            uint currentNum = atlasCount[tier].mul(ATLAS_NUMERATOR[tier]);
+
+            if (currentNum == denominator) {
```
Comment (krzysztof-jelski) [inline contracts/Lib/DmpAlgorithm.sol]: @Mike-S-Mike I think we can get rid of this `if` and leave only the part in `else`. Am I missing some special edge case or do you agrees?

Comment (Mike-S-Mike) [reply contracts/Lib/DmpAlgorithm.sol]: This case is needed if we only have one type of nodes, or if there are no nodes of a certain type. Without this we can get probability 99% (e.g. due to rounding errors) instead of 100. So, in that case if we roll 100 on the DMP, the algorithm won't be able to select anything and will fail.

Comment (krzysztof-jelski) [reply contracts/Lib/DmpAlgorithm.sol]: But then there's the correction for that case in lines 67-72, isn't it?

File: contracts/Configuration/Config.sol [hunk 20]
```diff
@@ -16,19 +16,16 @@ contract Config {
...
-
-    uint32 constant public COOLDOWN_TIMEOUT = 1 days;
 
     uint constant public APOLLO_DEPOSIT = 250000 ether;
-    uint constant public ATLAS1_STAKE = 10000 ether;
-    uint constant public ATLAS2_STAKE = 30000 ether;
-    uint constant public ATLAS3_STAKE = 75000 ether;
+
+    uint[3] public ATLAS_STAKE = [10000 ether, 30000 ether, 75000 ether];
+    uint[3] public ATLAS_NUMERATOR = [1, 2, 7];
```
[the remark this answers is not in the document]
Comment (Mike-S-Mike) [reply contracts/Configuration/Config.sol]: Fixed

File: contracts/Lib/DmpAlgorithm.sol (hunk 19 above)
Comment (Mike-S-Mike) [reply contracts/Lib/DmpAlgorithm.sol]: Yes, you are right, 67-72 lines have been added later than 'If' statement and they resolve such situation. 'If' statement was removed.
